# Capstone Project

### Version 7

Final version.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes
import folium

print('Libraries imported!')

Solving environment: done

## Package Plan ##

  environment location: /home/jupyterlab/conda/envs/python

  added / updated specs:
    - geopy


The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    geographiclib-1.50         |             py_0          34 KB  conda-forge
    geopy-1.22.0               |     pyh9f0ad1d_0          63 KB  conda-forge
    ------------------------------------------------------------
                                           Total:          97 KB

The following NEW packages will be INSTALLED:

  geographiclib      conda-forge/noarch::geographiclib-1.50-py_0
  geopy              conda-forge/noarch::geopy-1.22.0-pyh9f0ad1d_0



geopy-1.22.0         | 63 KB     | ##################################### | 100% 
geographiclib-1.50   | 34 KB     | ##################################### | 100% 
Preparing transaction: done
Verifying transaction: done
Executing transaction: 

In [2]:
pip install lxml

     |████████████████████████████████| 5.5MB 2.7MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.



### Scraping of table from specified URL

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
tables = pd.read_html(url, header=0,flavor=['lxml'])
# tables[0]

### Conversion of table to dataframe

In [4]:
df = pd.DataFrame(tables[0])

In [5]:
df.shape

(180, 3)

In [6]:
df.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

### Removal of rows with no assigned borough

In [7]:
df1 = df[df.Borough != 'Not assigned']

In [8]:
df1.shape

(103, 3)

In [9]:
df1.head()

Postal Code           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [10]:
df1.reset_index(drop=True).head()

Postal Code           Borough                                 Neighborhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### Checking for duplicate postal codes

In [11]:
len(df1['Postal Code'].unique())

103

##### The number of Postal Codes is equal to the number of rows of df1, this implies that there are no duplicates.

### Checking for missing neighborhoods

In [12]:
df1.count()

Postal Code     103
Borough         103
Neighborhood    103
dtype: int64

##### The number of values is equal to the number of rows in df1, this implies that here are no missing neighborhoods.

### Size of the resulting dataframe of neighborhoods

In [13]:
df1.shape

(103, 3)

##### The dataframe contains 103 rows.

### Reading the csv file with the Postal Code coordinates

In [14]:
df3 = pd.read_csv('Geospatial_Coordinates.csv')
df3.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [15]:
df3.shape

(103, 3)

In [16]:
df4 = pd.merge(df1, df3, how='left', on='Postal Code')

### Resulting dataframe with latitude and longitude

In [17]:
df4.head()

Postal Code           Borough                                 Neighborhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

### Exploration and clustering of neighborhoods in Toronto

In [18]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Here is a map of the area under study

In [19]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)
map_toronto

### Creation of new dataframe of Boroughs that contain Toronto

A new data frame with the Boroughs that contain the word Toronto was created in order to work 
with the neighborhoods because Folium does not provide venue information for some of the 
neighborhoods in the complete Toronto area. Therefore, from now on, the work shown is done
with the reduced number of Boroughs.

In [20]:
df5 = df4[df4['Borough'].str.contains('Toronto')].reset_index(drop=True)

In [21]:
df5 = df5.sort_values(by='Neighborhood')

In [22]:
df5.reset_index(drop=True).head()

Postal Code           Borough  \
0         M5E  Downtown Toronto   
1         M6K      West Toronto   
2         M7Y      East Toronto   
3         M5V  Downtown Toronto   
4         M5G  Downtown Toronto   

                                        Neighborhood   Latitude  Longitude  
0                                        Berczy Park  43.644771 -79.373306  
1       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
2  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
3  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
4                                 Central Bay Street  43.657952 -79.387383

In [23]:
df5.shape

(39, 5)

The work will be done with 39 neighborhoods.

In [24]:
latlon = df5[['Latitude','Longitude']].values

In [25]:
latlon[0:5,:]

array([[ 43.6447708, -79.3733064],
       [ 43.6368472, -79.4281914],
       [ 43.6627439, -79.321558 ],
       [ 43.6289467, -79.3944199],
       [ 43.6579524, -79.3873826]])

In [26]:
from math import cos, asin, sqrt, pi

def distance(lat1, lon1, lat2, lon2):
    p = pi/180
    a = 0.5 - cos((lat2-lat1)*p)/2 + cos(lat1*p) * cos(lat2*p) * (1-cos((lon2-lon1)*p))/2
    return 12742 * asin(sqrt(a)) #2*R*asin...

In [27]:
neigh = len(latlon)
distances = np.empty([39,39])
for i in range(0,neigh):
    for j in range(0,neigh):
        if i == j:
            distances[i,j] = np.inf
        else:
            lat1 = latlon[i,0]
            lon1 = latlon[i,1]
            lat2 = latlon[j,0]
            lon2 = latlon[j,1]
            distances[i,j] = distance(lat1, lon1, lat2, lon2)

In [28]:
dist_min = np.empty([39])
for i in range(0,neigh):
    dist_min[i] = min(distances[i,:])

In [29]:
dist_min

array([0.22270717, 1.40684384, 0.8459053 , 1.66766087, 0.51255249,
       1.40300209, 0.64508797, 0.18151811, 0.94381041, 0.94381041,
       1.58527051, 0.15033369, 1.48042065, 0.68508037, 0.70746365,
       1.56172505, 0.8459053 , 1.05522873, 1.70164277, 1.40684384,
       1.20179543, 0.99339106, 1.56172505, 0.51255249, 1.22839001,
       0.30099349, 1.20179543, 0.99339106, 1.9375388 , 0.50945114,
       1.2673442 , 0.22270717, 1.59834516, 1.40279269, 1.20204816,
       1.98897242, 1.7923045 , 0.15033369, 0.85022024])

In [30]:
min(dist_min)

0.1503336883953562

The minimum distance between neighborhoods is 150 m

In [31]:
dist_min < 0.5

array([ True, False, False, False, False, False, False,  True, False,
       False, False,  True, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True, False,
       False, False, False, False,  True, False, False, False, False,
       False,  True, False])

There are 5 neoghborhoods with a minimum distance of less than 500 m to other neighborhoods.

In [32]:
dist_ave = np.empty([39])
for i in range(0,neigh):
    distances[i,i] = 0
    dist_ave[i] = sum(distances[i,:])/(neigh-1)

In [33]:
from statistics import mean
mean(dist_ave)

4.944361252336709

The average distance between neighborhoods is 4.94 Km. I will stick to a radius of 500 m.

### Here is a map of the 39 neighborhoods

In [34]:
# Superimposing neighborhoods in part of Toronto
for lat, lng, borough, neighborhood in zip(df5['Latitude'], df5['Longitude'], df5['Borough'], df5['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### Beginning of analysis with Foursquare

In [35]:
CLIENT_ID = '3W1A2N3QRJCNIBMRQEHS5AX0P4JGOKFTZCCBOO0X5CRXCEFF' # your Foursquare ID
CLIENT_SECRET = 'AHW0EUIOGFD1TNT0BGOB3Q1ERYXRZHTC2XKGSIAHTGMZZRFZ' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 3W1A2N3QRJCNIBMRQEHS5AX0P4JGOKFTZCCBOO0X5CRXCEFF
CLIENT_SECRET:AHW0EUIOGFD1TNT0BGOB3Q1ERYXRZHTC2XKGSIAHTGMZZRFZ


### Exploring 39 neighborhoods in Toronto

In [36]:
def getNearbyVenues(names, latitudes, longitudes):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [37]:
LIMIT = 100
radius = 500
toronto_venues = getNearbyVenues(names=df5['Neighborhood'],
                                   latitudes=df5['Latitude'],
                                   longitudes=df5['Longitude']
                                  )

Berczy Park
Brockton, Parkdale Village, Exhibition Place
Business reply mail Processing Centre, South Central Letter Processing Plant Toronto
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Central Bay Street
Christie
Church and Wellesley
Commerce Court, Victoria Hotel
Davisville
Davisville North
Dufferin, Dovercourt Village
First Canadian Place, Underground city
Forest Hill North & West, Forest Hill Road Park
Garden District, Ryerson
Harbourfront East, Union Station, Toronto Islands
High Park, The Junction South
India Bazaar, The Beaches West
Kensington Market, Chinatown, Grange Park
Lawrence Park
Little Portugal, Trinity
Moore Park, Summerhill East
North Toronto West, Lawrence Park
Parkdale, Roncesvalles
Queen's Park, Ontario Provincial Government
Regent Park, Harbourfront
Richmond, Adelaide, King
Rosedale
Roselawn
Runnymede, Swansea
St. James Town
St. James Town, Cabbagetown
Stn A PO Boxes
Studio District
Summerhill West, Ra

In [38]:
print(toronto_venues.shape)
toronto_venues.head()

(1623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Berczy Park              43.644771              -79.373306   
1  Berczy Park              43.644771              -79.373306   
2  Berczy Park              43.644771              -79.373306   
3  Berczy Park              43.644771              -79.373306   
4  Berczy Park              43.644771              -79.373306   

                                       Venue  Venue Latitude  Venue Longitude  \
0                                       LCBO       43.642944       -79.372440   
1       The Keg Steakhouse + Bar - Esplanade       43.646712       -79.374768   
2                             Fresh On Front       43.647815       -79.374453   
3                              Meridian Hall       43.646292       -79.376022   
4  Hockey Hall Of Fame (Hockey Hall of Fame)       43.646974       -79.377323   

                  Venue Category  
0                   Liquor Store  
1                     Restaurant  
2  Vegetarian / Vegan Restaurant  
3                   Concert Hall  
4                         Museum

In [39]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     15   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     65   
Christie                                                               17   
Church and Wellesley                                                   81   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           15   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              61   
Lawrence Park                                                           3   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             4   
North Toronto West, Lawrence Park                                      19   
Parkdale, Roncesvalles                                                 15   
Queen's Park, Ontario Provincial Government                            33   
Regent Park, Harbourfront                                              45   
Richmond, Adelaide, King                                               93   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     37   
St. James Town                                                         80   
St. James Town, Cabbagetown                                            47   
Stn A PO Boxes                                                         97   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    21   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      15   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      65   
Christie                                                                17   
Church and Wellesley                                                    81   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

### Checking the total number of neighborhoods under study

In [40]:
len(toronto_venues['Neighborhood'].unique())

39

In [41]:
toronto_venues.shape

(1623, 7)

### Analysis of each neighborhood

In [42]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# rename column Neighborhood
toronto_onehot.rename(columns={"Neighborhood":"Neighbor"},inplace=True)

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood']

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]
toronto_onehot.head()

Neighborhood  Afghan Restaurant  Airport  Airport Food Court  \
0  Berczy Park                  0        0                   0   
1  Berczy Park                  0        0                   0   
2  Berczy Park                  0        0                   0   
3  Berczy Park                  0        0                   0   
4  Berczy Park                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  \
0                 0                   0              0          0           0   
1                 0                   0              0          0           0   
2                 0                   0              0          0           0   
3                 0                   0              0          0           0   
4                 0                   0              0          0           0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0           0       0     0    0                 0                   0      0   
1           0       0     0    0                 0                   0      0   
2           0       0     0    0                 0                   0      0   
3           0       0     0    0                 0                   0      0   
4           0       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
0        0     0                          0             0            0   
1        0     0                          0             0            0   
2        0     0                          0             0            0   
3        0     0                

In [43]:
toronto_onehot.shape

(1623, 237)

In [44]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').sum().reset_index()
toronto_grouped[0:5]

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                  0   
1       Brockton, Parkdale Village, Exhibition Place                  0   
2  Business reply mail Processing Centre, South C...                  0   
3  CN Tower, King and Spadina, Railway Lands, Har...                  0   
4                                 Central Bay Street                  0   

   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0        0                   0               0                0   
1        0                   0               0                0   
2        0                   0               0                0   
3        1                   1               2                3   
4        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            1   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 2                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           1                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          1           0           1       2     0    0   
1              0          0           0           0       1     0    1   
2              1          0           0           0       0     0    0   
3              0          0           0           0       0     0    1   
4              0          0           0           0       0     0    1   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   1      1                0         2   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       1              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              1          0         1   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               1        0                0         0   
1                     0               2        0                0         0   
2                     0               0        1                0         0   
3                     0               0        0                0         0   
4                     0               0        0                2         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        1     2   
1             0              1         0        0     3   
2             0              1         0        0     0   
3             0              0         0        0     0   
4             2              0         0        0     3   

   Cajun / Creole

In [45]:
toronto_grouped.set_index('Neighborhood', inplace=True)

### Number of venues per neighborhood

In [46]:
toronto_grouped.sum(axis=1)

Neighborhood
Berczy Park                                                                                                    57
Brockton, Parkdale Village, Exhibition Place                                                                   23
Business reply mail Processing Centre, South Central Letter Processing Plant Toronto                           15
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport     17
Central Bay Street                                                                                             65
Christie                                                                                                       17
Church and Wellesley                                                                                           81
Commerce Court, Victoria Hotel                                                                                100
Davisville                                                                 

In [47]:
toronto_grouped.shape

(39, 236)

### Procedure for gathering the 237 neighborhoods in 10 main neighborhoods

trans_mat = pd.read_csv('trans_matrix.csv',sep=';')
trans_mat.head()

trans_mat.shape

toronto_grouped.columns

trans_mat.set_index(toronto_grouped.columns, inplace=True)

trans_mat.head()

toronto_regrouped = toronto_grouped.dot(trans_mat)

toronto_regrouped.shape

### Loading of the resulting dataframe toronto_regrouped dataframe

In [60]:
toronto_regrouped = pd.read_csv('toronto_regrouped.csv')

In [61]:
toronto_regrouped.head()

Neighborhood  Arts & enterteinment  \
0                                        Berczy Park                     3   
1       Brockton, Parkdale Village, Exhibition Place                     1   
2  Business reply mail Processing Centre, South C...                     0   
3  CN Tower, King and Spadina, Railway Lands, Har...                     0   
4                                 Central Bay Street                     1   

   College & university  Event  Food  Nightlife spot  Outdoors & recreation  \
0                     0      0    25               8                      4   
1                     0      0    10               2                      1   
2                     0      0     4               1                      4   
3                     0      0     2               1                      1   
4                     0      0    49               3                      2   

   Professional & other places  Residence   Shop & service  Travel & transport  
0                            1           0              14                   1  
1                            3           0               6                   0  
2                            3           0               3                   1  
3                            0           0               1                  13  
4                            4           0               5                   1

In [62]:
toronto_regrouped.sum()

Neighborhood                   Berczy ParkBrockton, Parkdale Village, Exhibit...
Arts & enterteinment                                                          67
College & university                                                           4
Event                                                                          0
Food                                                                         864
Nightlife spot                                                               150
Outdoors & recreation                                                        101
Professional & other places                                                  101
Residence                                                                      0
Shop & service                                                               268
Travel & transport                                                            63
dtype: object

### Addition of location information to the toronto_grouped dataframe

In [63]:
df5.head()

Postal Code           Borough  \
5          M5E  Downtown Toronto   
14         M6K      West Toronto   
38         M7Y      East Toronto   
32         M5V  Downtown Toronto   
6          M5G  Downtown Toronto   

                                         Neighborhood   Latitude  Longitude  
5                                         Berczy Park  43.644771 -79.373306  
14       Brockton, Parkdale Village, Exhibition Place  43.636847 -79.428191  
38  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
32  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
6                                  Central Bay Street  43.657952 -79.387383

In [64]:
df6 = df5[['Neighborhood']]

In [65]:
df6['Latitude'] = 50 * (df5['Latitude'] - df5['Latitude'].min()) / (df5['Latitude'].max() - df5['Latitude'].min())
df6['Longitude'] = 50 * (df5['Longitude'] - df5['Longitude'].min()) / (df5['Longitude'].max() - df5['Longitude'].min())

/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [66]:
df6.head()

Neighborhood   Latitude  Longitude
5                                         Berczy Park   7.986016  29.031516
14       Brockton, Parkdale Village, Exhibition Place   3.987179  14.695142
38  Business reply mail Processing Centre, South C...  17.056578  42.548586
32  CN Tower, King and Spadina, Railway Lands, Har...   0.000000  23.516511
6                                  Central Bay Street  14.638431  25.354707

In [67]:
toronto_grouped = toronto_regrouped
toronto_grouped.head()

Neighborhood  Arts & enterteinment  \
0                                        Berczy Park                     3   
1       Brockton, Parkdale Village, Exhibition Place                     1   
2  Business reply mail Processing Centre, South C...                     0   
3  CN Tower, King and Spadina, Railway Lands, Har...                     0   
4                                 Central Bay Street                     1   

   College & university  Event  Food  Nightlife spot  Outdoors & recreation  \
0                     0      0    25               8                      4   
1                     0      0    10               2                      1   
2                     0      0     4               1                      4   
3                     0      0     2               1                      1   
4                     0      0    49               3                      2   

   Professional & other places  Residence   Shop & service  Travel & transport  
0                            1           0              14                   1  
1                            3           0               6                   0  
2                            3           0               3                   1  
3                            0           0               1                  13  
4                            4           0               5                   1

In [68]:
toronto_grouped = pd.merge(df6[['Neighborhood','Latitude','Longitude']], toronto_grouped, how='left', on='Neighborhood')
toronto_grouped.head()

Neighborhood   Latitude  Longitude  \
0                                        Berczy Park   7.986016  29.031516   
1       Brockton, Parkdale Village, Exhibition Place   3.987179  14.695142   
2  Business reply mail Processing Centre, South C...  17.056578  42.548586   
3  CN Tower, King and Spadina, Railway Lands, Har...   0.000000  23.516511   
4                                 Central Bay Street  14.638431  25.354707   

   Arts & enterteinment  College & university  Event  Food  Nightlife spot  \
0                     3                     0      0    25               8   
1                     1                     0      0    10               2   
2                     0                     0      0     4               1   
3                     0                     0      0     2               1   
4                     1                     0      0    49               3   

   Outdoors & recreation  Professional & other places  Residence   \
0                      4                            1           0   
1                      1                            3           0   
2                      4                            3           0   
3                      1                            0           0   
4                      2                            4           0   

   Shop & service  Travel & transport  
0              14                   1  
1               6                   0  
2               3                   1  
3               1                  13  
4               5                   1

In [69]:
toronto_grouped.shape

(39, 13)

### Cluster neighborhoods

In [70]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 4, 3, 4, 1, 4, 1, 1, 2, 2, 4, 1, 2, 1, 1, 4, 3, 0, 2, 0, 2, 2,
       4, 0, 0, 1, 2, 2, 4, 1, 0, 1, 0, 2, 4, 3, 0, 1, 0], dtype=int32)

This is a much better result of clustering than the one that was made with a radius of 500 m. In
this case, Kmeans was able to separate the neighborhoods in similar clusters.

In [71]:
# add clustering labels
toronto_grouped.drop('Latitude',1,inplace=True)
toronto_grouped.drop('Longitude',1,inplace=True)
toronto_grouped = pd.merge(df5[['Neighborhood','Latitude','Longitude']], toronto_grouped, how='left', on='Neighborhood')
toronto_merged = toronto_grouped
toronto_merged.insert(1, 'Cluster Labels', kmeans.labels_)

toronto_merged.head() # check the last columns!

Neighborhood  Cluster Labels  \
0                                        Berczy Park               0   
1       Brockton, Parkdale Village, Exhibition Place               4   
2  Business reply mail Processing Centre, South C...               3   
3  CN Tower, King and Spadina, Railway Lands, Har...               4   
4                                 Central Bay Street               1   

    Latitude  Longitude  Arts & enterteinment  College & university  Event  \
0  43.644771 -79.373306                     3                     0      0   
1  43.636847 -79.428191                     1                     0      0   
2  43.662744 -79.321558                     0                     0      0   
3  43.628947 -79.394420                     0                     0      0   
4  43.657952 -79.387383                     1                     0      0   

   Food  Nightlife spot  Outdoors & recreation  Professional & other places  \
0    25               8                      4                            1   
1    10               2                      1                            3   
2     4               1                      4                            3   
3     2               1                      1                            0   
4    49               3                      2                            4   

   Residence   Shop & service  Travel & transport  
0           0              14                   1  
1           0               6                   0  
2           0               3                   1  
3           0               1                  13  
4           0               5                   1

In [72]:
df_labels = toronto_merged.groupby('Cluster Labels').count()
df_labels[['Neighborhood']]

Neighborhood
Cluster Labels              
0                          9
1                         10
2                          9
3                          3
4                          8

### Here is a map of the different clusters

In [73]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Top ten venues for each neighborhood

In [74]:
toronto_grouped.drop('Latitude',1,inplace=True)
toronto_grouped.drop('Longitude',1,inplace=True)
toronto_grouped.drop('Cluster Labels',1,inplace=True)

In [75]:
toronto_grouped.head()

Neighborhood  Arts & enterteinment  \
0                                        Berczy Park                     3   
1       Brockton, Parkdale Village, Exhibition Place                     1   
2  Business reply mail Processing Centre, South C...                     0   
3  CN Tower, King and Spadina, Railway Lands, Har...                     0   
4                                 Central Bay Street                     1   

   College & university  Event  Food  Nightlife spot  Outdoors & recreation  \
0                     0      0    25               8                      4   
1                     0      0    10               2                      1   
2                     0      0     4               1                      4   
3                     0      0     2               1                      1   
4                     0      0    49               3                      2   

   Professional & other places  Residence   Shop & service  Travel & transport  
0                            1           0              14                   1  
1                            3           0               6                   0  
2                            3           0               3                   1  
3                            0           0               1                  13  
4                            4           0               5                   1

In [76]:
#toronto_grouped.insert(1, 'Latitude',df6['Latitude'])
#toronto_grouped.insert(1, 'Longitude',df6['Longitude'])
#toronto_merged.drop('Cluster Labels',1,inplace=True)

In [77]:
toronto_merged.head()

Neighborhood  Arts & enterteinment  \
0                                        Berczy Park                     3   
1       Brockton, Parkdale Village, Exhibition Place                     1   
2  Business reply mail Processing Centre, South C...                     0   
3  CN Tower, King and Spadina, Railway Lands, Har...                     0   
4                                 Central Bay Street                     1   

   College & university  Event  Food  Nightlife spot  Outdoors & recreation  \
0                     0      0    25               8                      4   
1                     0      0    10               2                      1   
2                     0      0     4               1                      4   
3                     0      0     2               1                      1   
4                     0      0    49               3                      2   

   Professional & other places  Residence   Shop & service  Travel & transport  
0                            1           0              14                   1  
1                            3           0               6                   0  
2                            3           0               3                   1  
3                            0           0               1                  13  
4                            4           0               5                   1

In [78]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [79]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood  1st Most Common Venue  \
0                                        Berczy Park                   Food   
1       Brockton, Parkdale Village, Exhibition Place                   Food   
2  Business reply mail Processing Centre, South C...  Outdoors & recreation   
3  CN Tower, King and Spadina, Railway Lands, Har...     Travel & transport   
4                                 Central Bay Street                   Food   

  2nd Most Common Venue        3rd Most Common Venue  \
0        Shop & service               Nightlife spot   
1        Shop & service  Professional & other places   
2                  Food               Shop & service   
3                  Food               Shop & service   
4        Shop & service  Professional & other places   

         4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0        Outdoors & recreation   Arts & enterteinment    Travel & transport   
1               Nightlife spot  Outdoors & recreation  Arts & enterteinment   
2  Professional & other places     Travel & transport        Nightlife spot   
3        Outdoors & recreation         Nightlife spot            Residence    
4               Nightlife spot  Outdoors & recreation    Travel & transport   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Professional & other places            Residence                  Event   
1           Travel & transport            Residence                  Event   
2                   Residence                  Event  College & university   
3  Professional & other places                 Event  College & university   
4         Arts & enterteinment            Residence                  Event   

  10th Most Common Venue  
0   College & university  
1   College & university  
2   Arts & enterteinment  
3   Arts & enterteinment  
4   College & university

### Cluster examination

In [80]:
toronto_merged = neighborhoods_venues_sorted
toronto_merged.insert(1, 'Cluster Labels', kmeans.labels_)
toronto_merged.head()

Neighborhood  Cluster Labels  \
0                                        Berczy Park               0   
1       Brockton, Parkdale Village, Exhibition Place               4   
2  Business reply mail Processing Centre, South C...               3   
3  CN Tower, King and Spadina, Railway Lands, Har...               4   
4                                 Central Bay Street               1   

   1st Most Common Venue 2nd Most Common Venue        3rd Most Common Venue  \
0                   Food        Shop & service               Nightlife spot   
1                   Food        Shop & service  Professional & other places   
2  Outdoors & recreation                  Food               Shop & service   
3     Travel & transport                  Food               Shop & service   
4                   Food        Shop & service  Professional & other places   

         4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
0        Outdoors & recreation   Arts & enterteinment    Travel & transport   
1               Nightlife spot  Outdoors & recreation  Arts & enterteinment   
2  Professional & other places     Travel & transport        Nightlife spot   
3        Outdoors & recreation         Nightlife spot            Residence    
4               Nightlife spot  Outdoors & recreation    Travel & transport   

         7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Professional & other places            Residence                  Event   
1           Travel & transport            Residence                  Event   
2                   Residence                  Event  College & university   
3  Professional & other places                 Event  College & university   
4         Arts & enterteinment            Residence                  Event   

  10th Most Common Venue  
0   College & university  
1   College & university  
2   Arts & enterteinment  
3   Arts & enterteinment  
4   College & university

Cluster 0. Central zone: shopping-recreation.

In [81]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[0] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
0                                   Berczy Park                  Food   
17    Kensington Market, Chinatown, Grange Park                  Food   
19                     Little Portugal, Trinity                  Food   
23  Queen's Park, Ontario Provincial Government                  Food   
24                    Regent Park, Harbourfront                  Food   
30                  St. James Town, Cabbagetown                  Food   
32                              Studio District                  Food   
36                 The Danforth West, Riverdale                  Food   
38               University of Toronto, Harbord                  Food   

          2nd Most Common Venue        3rd Most Common Venue  \
0                Shop & service               Nightlife spot   
17               Shop & service        Outdoors & recreation   
19               Nightlife spot               Shop & service   
23  Professional & other places         Arts & enterteinment   
24               Shop & service  Professional & other places   
30               Shop & service               Nightlife spot   
32               Shop & service               Nightlife spot   
36               Shop & service               Nightlife spot   
38               Nightlife spot               Shop & service   

          4th Most Common Venue        5th Most Common Venue  \
0         Outdoors & recreation         Arts & enterteinment   
17               Nightlife spot  Professional & other places   
19        Outdoors & recreation         Arts & enterteinment   
23               Shop & service               Nightlife spot   
24        Outdoors & recreation               Nightlife spot   
30        Outdoors & recreation  Professional & other places   
32  Professional & other places        Outdoors & recreation   
36  Professional & other places        Outdoors & recreation   
38  Professional & other places         College & university   

          6th Most Common Venue        7th Most Common Venue  \
0            Travel & transport  Professional & other places   
17           Travel & transport         Arts & enterteinment   
19  Professional & other places           Travel & transport   
23        Outdoors & recreation         College & university   
24         Arts & enterteinment           Travel & transport   
30         Arts & enterteinment           Travel & transport   
32           Travel & transport                   Residence    
36           Travel & transport                   Residence    
38         Arts & enterteinment           Travel & transport   

   8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0             Residence                   Event   College & university  
17            Residence                   Event   College & university  
19            Residence                   Event   College & university  
23    Travel & transport             Residence                   Event  
24            Residence                   Event   College & university  
30            Residence                   Event   College & university  
32                 Event   College & university   Arts & enterteinment  
36                 Event   College & university   Arts & enterteinment  
38            Residence   Outdoors & recreation                  Event

Cluster 1. Downtown: business-recreation.

In [82]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[0] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
4                                  Central Bay Street                  Food   
6                                Church and Wellesley                  Food   
7                      Commerce Court, Victoria Hotel                  Food   
11             First Canadian Place, Underground city                  Food   
13                           Garden District, Ryerson                  Food   
14  Harbourfront East, Union Station, Toronto Islands                  Food   
25                           Richmond, Adelaide, King                  Food   
29                                     St. James Town                  Food   
31                                     Stn A PO Boxes                  Food   
37           Toronto Dominion Centre, Design Exchange                  Food   

    2nd Most Common Venue        3rd Most Common Venue  \
4          Shop & service  Professional & other places   
6          Shop & service               Nightlife spot   
7          Nightlife spot               Shop & service   
11         Nightlife spot           Travel & transport   
13         Shop & service  Professional & other places   
14  Outdoors & recreation               Nightlife spot   
25         Shop & service  Professional & other places   
29         Shop & service               Nightlife spot   
31         Shop & service               Nightlife spot   
37     Travel & transport               Shop & service   

          4th Most Common Venue  5th Most Common Venue  \
4                Nightlife spot  Outdoors & recreation   
6   Professional & other places  Outdoors & recreation   
7   Professional & other places     Travel & transport   
11  Professional & other places   Arts & enterteinment   
13        Outdoors & recreation         Nightlife spot   
14               Shop & service   Arts & enterteinment   
25               Nightlife spot   Arts & enterteinment   
29  Professional & other places  Outdoors & recreation   
31  Professional & other places  Outdoors & recreation   
37               Nightlife spot   Arts & enterteinment   

          6th Most Common Venue        7th Most Common Venue  \
4            Travel & transport         Arts & enterteinment   
6            Travel & transport         Arts & enterteinment   
7         Outdoors & recreation         Arts & enterteinment   
11               Shop & service        Outdoors & recreation   
13         Arts & enterteinment           Travel & transport   
14           Travel & transport  Professional & other places   
25           Travel & transport        Outdoors & recreation   
29           Travel & transport         Arts & enterteinment   
31         Arts & enterteinment           Travel & transport   
37  Professional & other places        Outdoors & recreation   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4             Residence                  Event   College & university  
6             Residence                  Event   College & university  
7             Residence                  Event   College & university  
11            Residence                  Event   College & university  
13  College & university            Residence                   Event  
14            Residence                  Event   College & university  
25            Residence                  Event   College & university  
29            Residence                  Event   College & university  
31            Residence                  Event   College & university  
37            Residence                  Event   College & university

Cluster 2. Northen zone: Residential-recreational area.

In [83]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[0] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood  \
8                                          Davisville   
9                                    Davisville North   
12    Forest Hill North & West, Forest Hill Road Park   
18                                      Lawrence Park   
20                        Moore Park, Summerhill East   
21                  North Toronto West, Lawrence Park   
26                                           Rosedale   
27                                           Roselawn   
33  Summerhill West, Rathnelly, South Hill, Forest...   

          1st Most Common Venue        2nd Most Common Venue  \
8                          Food               Shop & service   
9                          Food               Shop & service   
12                         Food               Shop & service   
18           Travel & transport  Professional & other places   
20  Professional & other places        Outdoors & recreation   
21                         Food               Shop & service   
26        Outdoors & recreation           Travel & transport   
27               Shop & service        Outdoors & recreation   
33                         Food               Nightlife spot   

          3rd Most Common Venue  4th Most Common Venue  \
8   Professional & other places  Outdoors & recreation   
9   Professional & other places     Travel & transport   
12        Outdoors & recreation     Travel & transport   
18        Outdoors & recreation         Shop & service   
20           Travel & transport         Shop & service   
21  Professional & other places     Travel & transport   
26               Shop & service             Residence    
27                         Food     Travel & transport   
33           Travel & transport         Shop & service   

          5th Most Common Venue        6th Most Common Venue  \
8                Nightlife spot           Travel & transport   
9         Outdoors & recreation                   Residence    
12                   Residence   Professional & other places   
18                   Residence                Nightlife spot   
20                   Residence                Nightlife spot   
21        Outdoors & recreation                   Residence    
26  Professional & other places               Nightlife spot   
27                   Residence   Professional & other places   
33  Professional & other places                   Residence    

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
8              Residence                  Event  College & university   
9          Nightlife spot                 Event  College & university   
12         Nightlife spot                 Event  College & university   
18                   Food                 Event  College & university   
20                   Food                 Event  College & university   
21         Nightlife spot                 Event  College & university   
26                   Food                 Event  College & university   
27         Nightlife spot                 Event  College & university   
33  Outdoors & recreation                 Event  College & university   

   10th Most Common Venue  
8    Arts & enterteinment  
9    Arts & enterteinment  
12   Arts & enterteinment  
18   Arts & enterteinment  
20   Arts & enterteinment  
21   Arts & enterteinment  
26   Arts & enterteinment  
27   Arts & enterteinment  
33   Arts & enterteinment

Cluster 3. Eastern zone: Recreational area.

In [84]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[0] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood  1st Most Common Venue  \
2   Business reply mail Processing Centre, South C...  Outdoors & recreation   
16                     India Bazaar, The Beaches West                   Food   
35                                        The Beaches  Outdoors & recreation   

   2nd Most Common Venue        3rd Most Common Venue  \
2                   Food               Shop & service   
16        Shop & service  Professional & other places   
35        Shop & service               Nightlife spot   

          4th Most Common Venue  5th Most Common Venue  \
2   Professional & other places     Travel & transport   
16               Nightlife spot  Outdoors & recreation   
35           Travel & transport             Residence    

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
2                Nightlife spot            Residence                  Event   
16         Arts & enterteinment    Travel & transport            Residence    
35  Professional & other places                  Food                 Event   

   9th Most Common Venue 10th Most Common Venue  
2   College & university   Arts & enterteinment  
16                 Event   College & university  
35  College & university   Arts & enterteinment

Cluster 4. Western zone: Residential area.

In [85]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[0] + list(range(2, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
1        Brockton, Parkdale Village, Exhibition Place                  Food   
3   CN Tower, King and Spadina, Railway Lands, Har...    Travel & transport   
5                                            Christie                  Food   
10                       Dufferin, Dovercourt Village        Shop & service   
15                      High Park, The Junction South                  Food   
22                             Parkdale, Roncesvalles                  Food   
28                                 Runnymede, Swansea                  Food   
34                The Annex, North Midtown, Yorkville                  Food   

   2nd Most Common Venue        3rd Most Common Venue  \
1         Shop & service  Professional & other places   
3                   Food               Shop & service   
5         Shop & service        Outdoors & recreation   
10        Nightlife spot                         Food   
15        Shop & service        Outdoors & recreation   
22        Shop & service        Outdoors & recreation   
28        Shop & service  Professional & other places   
34        Shop & service        Outdoors & recreation   

          4th Most Common Venue  5th Most Common Venue 6th Most Common Venue  \
1                Nightlife spot  Outdoors & recreation  Arts & enterteinment   
3         Outdoors & recreation         Nightlife spot            Residence    
5                Nightlife spot     Travel & transport            Residence    
10  Professional & other places  Outdoors & recreation  Arts & enterteinment   
15               Nightlife spot   Arts & enterteinment    Travel & transport   
22               Nightlife spot   Arts & enterteinment    Travel & transport   
28               Nightlife spot   Arts & enterteinment    Travel & transport   
34               Nightlife spot   Arts & enterteinment    Travel & transport   

          7th Most Common Venue        8th Most Common Venue  \
1            Travel & transport                   Residence    
3   Professional & other places                        Event   
5   Professional & other places                        Event   
10           Travel & transport                   Residence    
15                   Residence   Professional & other places   
22                   Residence   Professional & other places   
28                   Residence         Outdoors & recreation   
34                   Residence   Professional & other places   

   9th Most Common Venue 10th Most Common Venue  
1                  Event   College & university  
3   College & university   Arts & enterteinment  
5   College & university   Arts & enterteinment  
10                 Event   College & university  
15                 Event   College & university  
22                 Event   College & university  
28                 Event   College & university  
34                 Event   College & university